librad/src/uri.rs

```diff
@@ -0,0 +1,535 @@
+            self.proto.nss(),
+            multibase::encode(Base::Base32Z, &self.id)
+        )?;
+
+        if !self.path.is_empty() {
+            write!(
+                f,
+                "/{}",
+                percent_encode(self.path.as_bytes(), PATH_PERCENT_ENCODE_SET)
+            )?;
+        }
+
+        Ok(())
+    }
+}
+
+pub mod rad_urn {
+    use super::*;
+
+    #[derive(Debug, Error)]
+    #[non_exhaustive]
+    pub enum ParseError {
+        #[error("Missing {0}")]
+        Missing(&'static str),
+
+        #[error("Invalid namespace identifier: {0}")]
+        InvalidNID(String),
+
+        #[error("Invalid protocol: {0}")]
+        InvalidProto(String),
+
+        #[error("Malformed path")]
+        Path(#[from] path::ParseError),
+
+        #[error("Must be UTF8")]
+        Utf8(#[from] Utf8Error),
+
+        #[error("Invalid encoding")]
+        Encoding(#[from] multibase::Error),
+
+        #[error("Invalid multihash")]
+        Hash(#[from] multihash::DecodeOwnedError),
+    }
+}
+
+impl FromStr for RadUrn {
+    type Err = rad_urn::ParseError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let mut components = s.split(':');
+
+        let nid = components.next().ok_or(Self::Err::Missing("namespace"))?;
+        if nid != "rad" {
+            return Err(Self::Err::InvalidNID(nid.to_string()));
+        }
+
+        let proto = components
+            .next()
+            .ok_or_else(|| Self::Err::Missing("protocol"))
+            .and_then(|proto| {
+                Protocol::from_nss(proto).ok_or_else(|| Self::Err::InvalidProto(proto.to_string()))
+            })?;
+
+        components
+            .next()
+            .ok_or_else(|| Self::Err::Missing("id and path"))
+            .and_then(|id_and_path| {
+                let decoded = percent_decode_str(id_and_path).decode_utf8()?;
+                let mut iter = decoded.splitn(2, '/');
+                let id = iter
+                    .next()
+                    .ok_or_else(|| Self::Err::Missing("id"))
+                    .and_then(|id| {
+                        multibase::decode(id)
+                            .map(|(_, bytes)| bytes)
+                            .map_err(|e| e.into())
+                    })
+                    .and_then(|bytes| Multihash::from_bytes(bytes).map_err(|e| e.into()))?;
+                let path = match iter.next() {
+                    None => Ok(Path::new()),
+                    Some(path) => Path::parse(path),
+                }?;
+
+                Ok(Self { id, proto, path })
+            })
+    }
+}
+
+/// A `RadUrl` is a URL with the scheme `rad://`.
+///
+/// The authority of a rad URL is a [`PeerId`], from which to retrieve the
+/// `radicle-link` repository and branch identified by [`RadUrn`].
+#[derive(Clone, Debug, PartialEq)]
+pub struct RadUrl {
+    authority: PeerId,
+    urn: RadUrn,
+}
+
+impl RadUrl {
+    // TODO: we should be able to open a `RadUrl` from local storage
```

Comment:
Are we tracking this work in a ticket?

Reply:
Ya well, this has been transplanted from #82

Reply:
Right, makes sense :+1: